### ISSUE TITLE
Headers containing links result in invalid HTML; nested `a` tags

### ISSUE THREAD

**Issue by ollie27** [azerupi/mdBook#221]
If a header contains a link like the following it will result in nested `a` tags which is [invalid](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).

```markdown
## Header with [link](https://example.com).
```

This results in:

```html
<a class="header" href="foo.html#header-with-a-hrefhttpsexamplecomlinka" name="header-with-a-hrefhttpsexamplecomlinka"><h2>Header with <a href="https://example.com/">link</a>.</h2></a>
```

A possible solution is to put the anchor at the start or end of the header rather than around the whole thing.

**Comment by azerupi** [comment]
I might be wrong but is this not a bug with Pulldown-cmark?

**Comment by ollie27** [comment]
No. pulldown-cmark creates the inner link but it's mdBook which naively surrounds the entire header in another `<a>` tag: https://github.com/azerupi/mdBook/blob/74e75d2cfbba79bfde31fbb43cc86df0671e4606/src/renderer/html_handlebars/hbs_renderer.rs#L273

**Comment by azerupi** [comment]
Ah right! Sorry, my bad. 😄 

So the thing is that we would like the header to be a link to itself so that user can click on the anchor and share the link directly. This can't work if the header is already a link obviously but I have no idea how we should handle that. Maybe add a special case for headers containing a link? I'm not sure.

**Comment by ollie27** [comment]
You can see what GitHub does for example: https://github.com/azerupi/mdBook/blob/master/README.md

**Comment by larry0x** [comment]
I encountered this issue today, with footnotes:

```markdown
### I am a title[^1]

[^1]: I am a footnote
```

The number 1 in the title isn't rendered properly as a `<sup>`. Additionally, in the footnote, the return link (the "↩") doesn't work.

I see #3100 attempted to fix this but was closed. I'm wondering why is that, and are you still open to contributions on this issue?
